### ISSUE TITLE
Options in vw readable file are different than the ones provided

### ISSUE THREAD

**Issue by xinofekuator** [VowpalWabbit/vowpal_wabbit#3148]
#### Describe the bug
The options found in vw readable file are different than the ones provided to vw from command line.

#### To Reproduce
1. Create file `train.dat` with following content:
```
1:1:0.5 | user_age:25
2:0:0.5 | user_age:25
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
1:1:0.5 | user_age:27
2:0:0.5 | user_age:21
2:0:0.5 | user_age:23
2:0:0.5 | user_age:56
2:1:0.5 | user_age:55
2:1:0.5 | user_age:36
1:1:0.5 | user_age:25
2:0:0.5 | user_age:25
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
```
2. Run this command: `vw -d train.dat -f /tmp/model.vw --cb 2 --invert_hash /tmp/bandit_readable.txt --epsilon 0.1 --l1 1e-05 --l2 1e-05 --save_resume`
3. Run same  command without `save_resume`: `vw -d train.dat -f /tmp/model_no_save.vw --cb 2 --invert_hash /tmp/bandit_readable_no_save.txt --epsilon 0.1 --l1 1e-05 --l2 1e-05`
4. Compare options from both output files:

When providing `save_resume` the arguments found here are different than the ones provided, so not sure if this is just an error in the readable file or if the wrong values are being used for training the model.
```
options: --cb 2 --cb_type dr --csoaa 2 --epsilon 0.100000001490116 --l1_state 0.000234659602589839 --l2_state 0.999765367272452
```

When not providing `save_resume` the regularisation arguments do not appear at all.
```
options: --cb 2 --cb_type dr --csoaa 2 --epsilon 0.100000001490116
```

NOTE: the same issue can be reproduced using `--readable_model` instead of `--invert_hash`

#### Expected behavior
I expect to get in the readable vw output the same `l1` and `l2` values as provided from command line. 

I also expect not to get epsilon argument as it is not used with `--cb` (and in case the semantics of `options` is what is being provided instead of what is being used then I would expect to see the value provided which is not the case).

#### Observed Behavior
Different `l1` and `l2` values than the ones provided by input. Also different epsilon value which should not appear as it is not used for anything.

#### Environment

MacOS Catalina with vw version:
```
❯ vw --version
8.10.2 (git commit: 32cd2f33a)
```

EDIT: this also happens with `8.10.1 (git commit: 3887696)`

#### Additional context
Maybe related with https://github.com/VowpalWabbit/vowpal_wabbit/issues/1573

**Comment by jackgerrits** [comment]
A key things to point out before I get further is the options `l1` and `l1_state` are distinct options and represent different things. `l1` is not stored in a `save_resume` model but `l1_state` is added as part of the extra training state.

Not every command line options is saved into the model and is based on if an option is required to make predictions. `save_resume` extends this to options required to resume training.

`l1` and `l2` are tricky ones and the decision to include or exclude from `save_resume` is not a clear one. I asked John and he said that `l1` is not stored in the model as they can be damaging to model accuracy.  `l1_state` is saved to avoid rounding inconsistencies between a `save_resume` model and training on conjoined datasets.

If you wish to use l1 or l2 regularization that must be explicitly specified for each training run.

`--epsilon` is a `keep` option which means it is required for predictions and is therefore saved with a model even without `save_resume`. You're right though it is not relevant in the command line you supplied but the keep/options framework is not able to distinguish this and so any keep option is saved in the model.

So, overall I **think** things are working as expected here.

**Comment by xinofekuator** [comment]
> A key things to point out before I get further is the options l1 and l1_state are distinct options and represent different things. l1 is not stored in a save_resume model but l1_state is added as part of the extra training state.

I think that explains it, that was a misunderstanding on my side. Thanks for clarifying.

> l1 and l2 are tricky ones and the decision to include or exclude from save_resume is not a clear one. I asked John and he said that l1 is not stored in the model as they can be damaging to model accuracy.

So is the recommended way to use regularisation for training the model but not using it for making the predictions?

> --epsilon is a keep option which means it is required for predictions and is therefore saved with a model

Why the value has different decimals than the one provided from command line? Actually it seems that epsilon does influence the predictions (although I guess it is not supposed to be used with `cb`):

```
> echo "| user_age:25" | vw --initial_regressor /tmp/model.vw --cb 2 --epsilon 0.0 --l1 1e-05 --l2 1e-05 -t --quiet -p /dev/stdout
0.000000 1.000000
```

```
❯ echo "| user_age:25" | vw --initial_regressor /tmp/model.vw --cb 2 --epsilon 0.1 --l1 1e-05 --l2 1e-05 -t --quiet -p /dev/stdout
0.050000 0.950000
```

**Comment by jackgerrits** [comment]
> So is the recommended way to use regularisation for training the model but not using it for making the predictions?

Regularization is part of training and does not affect predictions directly. (Obviously other than indirectly during the training process)

> Why the value has different decimals than the one provided from command line? Actually it seems that epsilon does influence the predictions (although I guess it is not supposed to be used with cb):

`--epsilon 0.1` vs `--epsilon 0.100000001490116` is just due to float rounding errors. 

What is the command line that created that model? (You can dump it with `vw --initial_regressor /tmp/model.vw --readable_model readable.txt --no_stdin`) It looks like it was different to the original question you asked. It looks to me that it was produced using `--cb_explore 2`

If it were just `--cb` or you run that same command omitting the loaded model (`--initial_regressor /tmp/model.vw`) , then the predictions you see will just be the cb class (1 or 2). Then epsilon will not affect it. Since it looks like the model contains `--cb_explore` the epsilon parameter is affecting the predictions in the way I would expect.

**Comment by xinofekuator** [comment]
Yes, you are right. I had the incorrect assumption that the `--cb` argument would override the saved model `--cb_explore` or that at least I would get a warning/error. It is a bit confusing to me that the argument is ignored without notifying the user.

> Regularisation is part of training and does not affect predictions directly. (Obviously other than indirectly during the training process)

Yes, I was referring to affecting to the training process, but I guess that if you train with regularisation the correct thing to do is to predict with regularisation too. So my question is about if using regularisation with vw is not recommended at all. If it can be damaging to model accuracy I agree it is ok not having it by default, but if it is an argument provided by the user and in some cases it might help I am of the opinion that then it should be saved in the model. If regularisation never works, wouldn't then be better to just removed those options from vw?

Thanks for all the answers.

**Comment by jackgerrits** [comment]
Hm, that is confusing. At this moment I am not sure of a good way to warn of that situation but I will keep it in mind. This is because by definition `--cb_explore` implies `--cb` so also saying `--cb` does not clash. It agrees.

As far as I am aware, predicting with regularization is not a thing that is possible. Regularization is part of the training process, not part of making a prediction. You would use regularization to train the model then use that subsequent model to make a prediction.

I think what John meant is that keeping regularization turned on in the model forever can be a surprising default and if it isn't known can be damaging to the model. Regularization certainly works, is very helpful and recommended if you have a situation where it would benefit (such as removing no longer used weights from a model). Sorry if my answer was a bit confusing there. So I think the dangerous part was automatically saving regularization and not using it in general.

**Comment by xinofekuator** [comment]
Yes, sorry, the predict with regularisation part was not correct and I think I did not provide all the information in my answer, as I had in mind an use case where model is saved for daily trainings, so after first training as regularisation is removed and not remembered if you forget to pass those argument in next days trainings then the effect of regularisation that happened in the first training will be small. Although I guess this depends on the learning rate decay.

I get what you mean now and it makes total sense. I will close this issue then as everything is working as expected.

Thanks again for all the clear explanations.

**Comment by jackgerrits** [comment]
You're welcome! Glad I could clear things up

**Comment by xinofekuator** [comment]
> Hm, that is confusing. At this moment I am not sure of a good way to warn of that situation but I will keep it in mind. This is because by definition --cb_explore implies --cb so also saying --cb does not clash. It agrees.

I was doing some tests regarding this (in 8.10.2 due to this fixed bug https://github.com/VowpalWabbit/vowpal_wabbit/issues/3062) to see how it worked and this also means that once you start training with `cb_explore` with `--save_resume` you can not change that value in future trainings unless you restart the training process.

E.g. `vw -d train.dat -f /tmp/model.vw --cb_explore 2 --invert_hash /tmp/bandit_readable.txt --epsilon 0.1 --l1 1e-05 --l2 1e-05 --save_resume` and then run `vw -d train.dat -f /tmp/model.vw --cb 2 --invert_hash /tmp/bandit_readable.txt --epsilon 0.1 --l1 1e-05 --l2 1e-05 --save_resume --initial_regressor /tmp/model.vw`.

When making predictions it uses exploration mode: `echo "| user_age:25" | vw --initial_regressor /tmp/model.vw --cb 2 --epsilon 0.0 --l1 1e-05 --l2 1e-05 -t --quiet -p /dev/stdout`

Is the `--cb` also ignored during training? Not sure why someone would want to train with exploration though, so maybe `cb_explore` for training should not be allowed? (maybe there are use cases for it, not sure 😅 )

**Comment by pmineiro** [comment]
It is natural to assume that a policy could be trained and then exploration "turned on" after the fact.  The default off-policy learning algorithm and exploration strategy do work this way, for instance.  But we don't support this, because sometimes the exploration strategy dictates the off-policy learning strategy (e.g., softmax).

**Comment by xinofekuator** [comment]
> It is natural to assume that a policy could be trained and then exploration "turned on" after the fact. The default off-policy learning algorithm and exploration strategy could work this way, for instance. But we don't support this, because sometimes the exploration strategy dictates the off-policy learning strategy (e.g., softmax).

Not sure if I understood what is not supported. Training with `cb` and turning exploration for predicting is supported. 
I guess you meant turning exploration for training, right? That yields an error `Model content is corrupted`. 

Anyway my example was the opposite case where you start training with exploration and then switch training to `cb` which does not give an error but  then only allows you to predict with exploration only (`cb` is ignored in my command above).

> Training with explore on is harmless. 

Why harmless? Do you mean that `vw` ignores the exploration for training? Or that in general is harmless? I think that if you have an agent that does exploration that could impact the training process, but I could be wrong.

> We are working on exposing all layers of the reduction stack as outputs which will fix this problem.

That is great! Thanks! 😄
